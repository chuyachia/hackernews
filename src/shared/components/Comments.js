import React from "react";
import axios from "axios";
import styled from 'styled-components';

const UnstyledUl = styled.ul`
    list-style:none;
    margin: 0;
    padding: 0;
`;

const Author=styled.li`
    color:grey;
`;

const Comment = styled.div`
    border-left:1px dotted grey;
    padding-left:1rem;
    margin: 1rem
`;

const hnUrl= "https://hacker-news.firebaseio.com/v0/";
class Comments extends React.Component {
    constructor(props){
        super(props);
        this.state={
            replies:[]
        };
    }
    componentDidMount(){
        if(this.props.data.kids&&this.state.replies.length==0){
            var promises = this.props.data.kids.map(function(id){
                return axios.get(hnUrl+'item/'+id+'.json');
            });
            axios.all(promises)
            .then((results)=>{
                results = results.slice(0,21);
                this.setState({
                    replies :results.map(function(c,i){return <Comments key={i} data={c.data}/>})
                });
            });
        }
    }
    render(){
        return(
            <Comment>
                <Author>{this.props.data.by}&nbsp;said</Author>
                <li dangerouslySetInnerHTML={{__html: this.props.data.text}}></li>
                {this.props.data.kids&&
                <UnstyledUl>
                    {this.state.replies}
                </UnstyledUl>
                }
            </Comment>
         );
    }
}

export default Comments;