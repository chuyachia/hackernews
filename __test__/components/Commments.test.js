import item from '../../__mocks__/itemwkids.json';
import { shallow } from 'enzyme';
import React from 'react';
import Comments from '../../src/shared/components/Comments';
import * as enzyme from 'enzyme';
import ReactSixteenAdapter  from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

enzyme.configure({ adapter: new ReactSixteenAdapter() });


describe('Comments component test',()=>{
    let wrapper;
    beforeEach(()=>{wrapper = shallow(<Comments key={0} data={item}/>);})
    it('Comments component snapshot', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('Comments component has the same number of comments child as the number of kids in data', () => {
      expect(wrapper.find(Comments)).toHaveLength(item.kids.length);
    });
})