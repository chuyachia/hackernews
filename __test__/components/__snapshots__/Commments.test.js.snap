// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments component test Comments component snapshot 1`] = `
<styled.div>
  <styled.li>
    xxx
    Â said
  </styled.li>
  <li
    dangerouslySetInnerHTML={
      Object {
        "__html": "Fake comment",
      }
    }
  />
  <styled.ul>
    <Comments
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="0"
    />
    <Comments
      data={
        Object {
          "by": "xxx",
          "descendants": 71,
          "id": 1,
          "score": 111,
          "text": "Fake comment",
          "time": 1175714200,
          "title": "Fake title",
          "type": "story",
          "url": "http://www.getdropbox.com/u/2/screencast.html",
        }
      }
      key="1"
    />
  </styled.ul>
</styled.div>
`;
